gadgets/cmr/config/LandmarkDetection/gadgetron_cmr_landmark_detection_util.py

```diff
@@ -1,28 +1,115 @@
+"""Utility file for cmr landmark detection"""
+
+import os
 import sys
-import torch
-import torch.nn as nn
-import torch.optim as optim
-import torch.nn.functional as F
+import urllib.request
+import hashlib
+import fcntl
 import numpy as np
-import scipy
 from scipy import ndimage
 
-def cpad_2d(data, RO, E1):
-    '''
-    data: [dRO, dE1, N], padded it round center to [RO, E1, N]
-    return value: (s_ro, s_e1), starting point of data in padded array
-    '''
+class model_file_lock:
+    """A python inter-process file lock
+    
+    Adaptive from https://stackoverflow.com/questions/6931342/system-wide-mutex-in-python-on-linux
+    """
+    
+    def __init__(self, model_dest):
+        self.model_dest = model_dest
+    
+    def __enter__ (self):
+        self.fp = open(os.path.join(self.model_dest, "ai_model_lockfile.lck"), 'w')
+        fcntl.flock(self.fp.fileno(), fcntl.LOCK_EX)
+
+    def __exit__ (self, _type, value, tb):
+        fcntl.flock(self.fp.fileno(), fcntl.LOCK_UN)
+        self.fp.close()
+            
+def is_model_valid(model_file, model_sha256):
+    """Check whether the sha256 hash is valid"""
     
+    if not os.path.isfile(model_file):
+        return False
+
+    with open(model_file, "rb") as f:
+        file_hash = hashlib.sha256()
+        while chunk := f.read(65536):
+            file_hash.update(chunk)
+        
+    return model_sha256 == file_hash.hexdigest()
+
+
+def check_and_get_model(model_host, model_file, model_dest, model_sha256):
+    """Get the model from the storage account.
+    
+    Input:
+        model_host [str] : host address to download the model
+        model_file [str] : model file name
+        model_dest [str] : destination path to store the model
+        model_sha256 [str] : model sha256 hash
+    
+    Return:
+        0 or 1 for success or fail
+        
+    If the model already exist, check its integrity; if invalid, download file
+    If model cannot be downloaded, an exception is thrown
+    """
     try:
-        dRO, dE1, N = data.shape
-    except:
-        data = np.expand_dims(data, axis=2)
-        dRO, dE1, N = data.shape
+        print(f"check_and_get_model,  model_host is {model_host}")
+        print(f"check_and_get_model,  model_file is {model_file}")
+        print(f"check_and_get_model,  model_dest is {model_dest}")
+        print(f"check_and_get_model,  model_sha256 is {model_sha256}")
+        
+        # assemble the source and destination of model
+        model_url = model_host + model_file
+        destination = os.path.join(model_dest, model_file)
+        
+        print(f"check_and_get_model,  model_url is {model_url}")
+        print(f"check_and_get_model,  destination is {destination}")
+        
+        # make sure destination exist
+        os.makedirs(model_dest, exist_ok=True)
+        
+        # download the model if not exist
```

Comment:
I think you have to hold the lock for the entire duration of downloading and checking the file to avoid that another process tinkers with it while you check it.

Reply:
Lock is now holding entire duration.